pkg/services/datasources/service/datasource.go

```diff
@@ -205,10 +210,48 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
 		cmd.Name = getAvailableName(cmd.Type, dataSources)
 	}
 
-	if err := s.validateFields(ctx, cmd.Name, cmd.URL, cmd.Type, cmd.APIVersion); err != nil {
+	// Validate the command
+	jd, err := cmd.JsonData.ToDB()
+	if err != nil {
+		return nil, fmt.Errorf("invalid jsonData")
+	}
+
+	settings, err := s.mutate(ctx, backend.PluginContext{
+		OrgID:    cmd.OrgID,
+		PluginID: cmd.Type,
+	}, &backend.DataSourceInstanceSettings{
+		UID:                     cmd.UID,
+		Name:                    cmd.Name,
+		URL:                     cmd.URL,
+		Database:                cmd.Database,
+		JSONData:                jd,
+		DecryptedSecureJSONData: cmd.SecureJsonData,
+		Type:                    cmd.Type,
+		User:                    cmd.User,
+		BasicAuthEnabled:        cmd.BasicAuth,
+		BasicAuthUser:           cmd.BasicAuthUser,
+		APIVersion:              cmd.APIVersion,
+	})
+	if err != nil {
 		return nil, err
 	}
 
+	// The mutable properties
+	cmd.URL = settings.URL
```

Comment:
nit: for the sake of readability, it would be nice to only do this if we know settings have been mutated.

Reply:
Do you mean something like:
```
if cmd.URL != settings.URL {
  cmd.URL = settings.URL
}
```
I have no idea if that is better/cleaner/faster or anything -- if important, I can change (seems like more repetitive words to me); but this whole section will likely be revisited a bunch before we are happy with it.  In particular as we try to apply static validation and perhaps https://github.com/grafana/grafana/pull/88036

Reply:
no, I was thinking on something like the `mutate` function to return a boolean if the settings has been really mutated (e.g. it would return `false` if it doesn't have an apiVersion). Then here do:

```
if mutated {
  cmd.URL = settings.URL
  ...
```

But as I said, just a nit, feel free to ignore.